// schematica/dist/testing.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var Schematica_1 = __importDefault(require("./Schematica"));
var json = new Schematica_1.default();
var sch = json.createSchema({
    type: "object",
    required: "all",
    name: "user",
    properties: {
        id: "number",
        name: { type: "string", maxLength: 42, minLength: 3 },
        email: "string",
        password: "string"
    },
    strict: true,
});
var validator = json.buildValidator(sch, { errors: true });
var encoder = json.buildSerializer(sch, { onAdditionalProperty: "error" });
var parser = json.buildParser(sch);
var normalizer = json.buildNormalizer(sch);
var data = {
    id: 1,
    name: "Kronsy",
    email: "john@doe.com",
    password: "abc123",
};
var encoded = JSON.stringify(data);
console.log(validator(data));
console.log(encoder(data));
console.log(normalizer(data));
console.log(parser(encoded));
while (true) { }
