// inside-in/dist/Nestable.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
const kIsRoot = Symbol("Is Root?");
const kChildren = Symbol("Children");
const kParent = Symbol("Parent");
class Nestable {
    constructor() {
        this[_a] = [];
        this[kIsRoot] = true;
    }
    addChild(child) {
        if (child === this) {
            throw new Error("Nestable cannot be it's own child");
        }
        if (!child[kParent]) {
            child[kParent] = this;
            child[kIsRoot] = false;
        }
        this[kChildren].push(child);
    }
    getChildren() {
        return this[kChildren];
    }
    deepGetChildren() {
        const children = [];
        this.getChildren().forEach(child => {
            children.push(child);
            children.push(...child.deepGetChildren());
        });
        return children;
    }
    getRootNode() {
        if (this[kIsRoot]) {
            return this;
        }
        else if (this[kParent]) {
            return this[kParent].getRootNode();
        }
        else {
            throw new Error("Cannot Get Root Node For Nestable");
        }
    }
    getParent() {
        if (this[kParent]) {
            return this[kParent];
        }
        else {
            return undefined;
        }
    }
}
exports.default = Nestable;
_a = kChildren;
