// schematica/dist/Schematica.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
var Validator_1 = __importDefault(require("./lib/Validator/Validator"));
var ERR_UNKNOWN_REF_1 = __importDefault(require("./errors/Schematica/ERR_UNKNOWN_REF"));
var Encoder_1 = __importDefault(require("./lib/Encoder/Encoder"));
var _1 = require(".");
var Schemas_1 = __importDefault(require("./Schemas"));
var Normalizer_1 = __importDefault(require("./lib/Normalizer/Normalizer"));
var kValidator = Symbol("Validator");
var kParser = Symbol("Parser");
var kEncoder = Symbol("Encoder");
var kNormalizer = Symbol("Normalizer");
var kSchemaRefStore = Symbol("Schema Store");
var Schematica = (function () {
    function Schematica(options) {
        if (options === void 0) { options = {}; }
        this[_a] = new Map();
        this[kValidator] = new Validator_1.default();
        this[kEncoder] = new Encoder_1.default({ validator: this[kValidator] });
        this[kParser] = new _1.Parser({ validator: this[kValidator] });
        this[kNormalizer] = new Normalizer_1.default({ validator: this[kValidator] });
    }
    Schematica.prototype.getSchema = function (ref) {
        var schema = this[kSchemaRefStore].get(ref);
        if (schema) {
            return schema;
        }
        else {
            throw new ERR_UNKNOWN_REF_1.default();
        }
    };
    Schematica.prototype.buildValidator = function (schema, options) {
        return this[kValidator].build(schema, options);
    };
    Schematica.prototype.buildSerializer = function (schema, options) {
        if (options === void 0) { options = {}; }
        return this[kEncoder].build(schema, options);
    };
    Schematica.prototype.buildNormalizer = function (schema) {
        return this[kNormalizer].build(schema);
    };
    Schematica.prototype.createSchema = function (schema) {
        var sch = (0, Schemas_1.default)(schema, this[kSchemaRefStore]);
        if (sch.ref) {
            this[kSchemaRefStore].set(sch.ref, sch);
        }
        return sch;
    };
    Schematica.prototype.buildParser = function (schema) {
        return this[kParser].build(schema);
    };
    return Schematica;
}());
exports.default = Schematica;
_a = kSchemaRefStore;
